server/data/data-jmap/src/test/java/org/apache/james/jmap/api/preview/MessagePreviewStoreContract.java

```diff
@@ -114,6 +120,47 @@ default void storeShouldBeIdempotent() {
             .isEqualTo(PREVIEW_1);
     }
 
+    @Test
+    default void concurrentStoreShouldOverrideOldValueWhenSameMessageId() throws Exception {
+        int threadCount = 10;
+        int stepCount = 100;
+
+        ConcurrentTestRunner.builder()
+            .operation((thread, step) -> Mono.from(testee()
+                .store(TestMessageId.of(thread), Preview.from(String.valueOf(step))))
+                .block())
+            .threadCount(threadCount)
+            .operationCount(stepCount)
+            .runSuccessfullyWithin(Duration.ofMinutes(1));
+
+        IntStream.range(0, threadCount)
+            .forEach(index -> assertThat(Mono.from(testee()
+                    .retrieve(TestMessageId.of(index)))
+                    .block())
+                .isEqualTo(Preview.from(String.valueOf(stepCount - 1))));
+    }
+
+    @Test
+    default void storeShouldBeConsistentUponSingleKeyOperation() throws Exception {
+        int threadCount = 10;
+        int operationCount = 100;
+
+        ConcurrentTestRunner.builder()
+            .operation((thread, step) -> Mono.from(testee()
+                .store(MESSAGE_ID_1, Preview.from(String.valueOf(step * threadCount + thread))))
+                .block())
+            .threadCount(threadCount)
+            .operationCount(operationCount)
+            .runSuccessfullyWithin(Duration.ofMinutes(1));
+
+        Preview preview = Mono.from(testee().retrieve(MESSAGE_ID_1)).block();
+        Integer previewAsInt = Integer.valueOf(preview.getValue());
+
+        assertThat(previewAsInt)
+            .describedAs("Ensure the stored result was generated by the last operation of one of the threads")
```

Comment:
I don't see why it should be true

Reply:
Assuming there are 10 threads starting store previews with a same keys, There is no guarantee about thread orders, which one will be last, but we know the each last preview stored by each thread. With 100 operations each thread, the last storing preview should be one in the range 990(last of thread 0) -> 999(last of thread 9).

That's the best assertion we could do regarding this test. WDYT?

Reply:
take 2 threads T1 and T2 with 10 operations each.
You can have (time is going horizontally):
```
T1 op1 op2 op3 op4 op5 op6 op7 op8 op9 op10
T2  op1 op2 op3 op4 op5 op6 op7 op8 op9 op10
```
but you could also have:
```
T1 op1 op2 op3 op4 op5 op6 op7 op8 op9 op10
T2  op1  op2  op3  op4  op5  op6   op7   op8   op9 op10
```
Or do I miss something?

The property we have are:
* if a key doesn't have a value and we have concurrent store calls of a value V, we want to end with the value V for this key
* if a key have a value V1 and we have concurrent store calls of a value V2, we want to end with the value V2 for this key

Reply:
I think you didn't really get my scenario, Okay with 2 threads and 10 operations each, storing operation is described as a string "MessageId-Preview", e.g. "1-0" or "1-9", my runner model will be like:
```
T1 1-0    1-2     1-4   1-6     1-8      1-10    1-12  1-14   1-16             1-18
T2 1-1  1-3    1-5        1-7       1-9      1-11     1-13     1-15     1-17 1-19
```
The rule is:
 - each thread store only one key, concurrent is true from multiple-keys perspective and false from single key perspective
 - each thread will store incremented value time by time and the last value store should be the maximum value that thread had been stored.

=> So we don't know which thread will be last, but for a certain thing, the value is 18 or 19. And 18 -> 19 is the range I made from the assertion computation `threadCount * (operationCount - 1) -> threadCount * operationCount`

The properties you described looks like idempotent test on concurrent envirointment right? I store the same record concurrently and I expect the result is idempotent?

Reply:
In case of you having concerns about this, can I remove this test and resolving your concerns in another PR? 

I want to merge the PR soon be cause there are many depend on it. In case that problem is your only concern?

Reply:
We decided to merge this, as so we can advance in the sprint. However if you still have a concern with that @mbaechler , please comment and we will fix it in an other PR. Thanks !

Reply:
I'm not sure trying to test that specific behavior worth the investment.

I just don't want flaky tests to be introduced on broken expectation.

If I take @trantienduchn example, these are the things you don't know:
* how threads will be scheduled
* how long will the write take

So you can very well have the T2 thread taking 1-17, 1-18 and 1-19 and have 1-16 be the winner because T1 finally wrote it after T2 wrote 1-19.

Reply:
But T-1 will keep writing value until it writes the last value of T-1 whic is in the expected range.

Reply:
T-1 only write evens, T-2 only write odds, each thread store a small number, only after finishes a current number, then come to a bigger number... 

1-16 from thread T1 can very well stored after 1-19 from T2 but 1-16 always comes after 1-18 (because 1-16 and 1-18 are executed by same thread, and it follow the rule of the incrementation)

Reply:
Ok, I get it.
So please make the code say that (because it doesn't) and I'll be ok.